### PR TITLE
more changes

### PR DESCRIPTION
**Pull request by rigazilla**
- Changed protobuf message from `Application` to `AppEntry`
- added more info on Marshaller, protostream, interoperability

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,18 @@
 # Infinispan Hot Rod .NET Core Demo
 
 This branch provides a demonstration of the Hot Rod .NET Core client with remote caches on an Infinispan Server instance.
+### Goal
+This demo aims to show:
+- the current maturity level of the new .NET Core client library;
+- how to include it in your application;
+- how to store and query entries in remote cache;
+- how the client handles Protobuf serialization.
```
Comment (oraNod) [inline README.md]:
```suggestion
This demo aims to show you the current maturity level of the Hot Rod .NET Core client as well as how to:

- Include ithe client in your application.
- Store and query entries in a remote cache.
- Handle Protobuf serialization of C# data types.
```

File: README.md [hunk 2]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
 
 ### Some highlights
 
-- .NET applications can use plain C# data types. By configuring Infinispan caches with the ProtoStream marshaller the Hot Rod .NET Core client handles all the Protobuf to C# conversion ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/e2efac6591741d23ff92c6253bf1257a60ea8879/demo/Query/Program.cs#L35-L38)).
-- The Hot Rod client API is asynchronous by default and concurrency is supported. This demo runs all PUT commands in a single call ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/e2efac6591741d23ff92c6253bf1257a60ea8879/demo/Query/Program.cs#L111-L122)).
+- Concurrency: the Hot Rod client API is asynchronous by default and concurrency is supported. To show that, the demo runs all PUT commands asynchronously ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/e2efac6591741d23ff92c6253bf1257a60ea8879/demo/Query/Program.cs#L111-L122)).
```
Comment (oraNod) [inline README.md]:
```suggestion
- Concurrency: The Hot Rod client API is asynchronous by default and supports concurrency. To show these capabilites the demo runs all PUT commands asynchronously ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/e2efac6591741d23ff92c6253bf1257a60ea8879/demo/Query/Program.cs#L111-L122)).
```

File: README.md [hunk 3]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
+
+- Simplicity: the client is shipped as a .nuget package and can be easily imported ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Query.csproj#L7)).
```
Comment (oraNod) [inline README.md]:
```suggestion
- Simplicity: The client is shipped as a `.nuget` package that you can easily import ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Query.csproj#L7)).
```

File: README.md [hunk 4]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
+- Protobuf support: using the default Protobuf/Grpc tools ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Query.csproj#L8-L9)), the C# data types hierarchy can be generated and used in the application.
```
Comment (oraNod) [inline README.md]:
```suggestion
- Protobuf support: Using the default Protobuf/GRPC tools ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Query.csproj#L8-L9)), the C# data types hierarchy can be generated and used in the application.
```

File: README.md [hunk 5]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
 
-To query remote caches you need to provide Infinispan Server with Protobuf schema that describe your data model ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/e2efac6591741d23ff92c6253bf1257a60ea8879/demo/Query/Program.cs#L69)).
+#### Client Marshaller
+Infinispan uses `application/x-protostream` media type to serialize remote query. [Protostream](https://github.com/infinispan/protostream) is a Protobuf based protocol that allows multiple types of object to be serialized onto the same data stream. Simplifying a little bit: client and server share a map that assigns an integer number to each queriable object type; this is usually done defining the `TypeId` in the `.proto` file. (see [link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/app.proto#L5) and [link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/review.proto#L5)). This typeId is sent in front of each Protobuf message sent on the wire.
```
Comment (oraNod) [inline README.md]:
```suggestion
Infinispan uses the `application/x-protostream` media type to serialize data for remote queries. [ProtoStream](https://github.com/infinispan/protostream) is a Protobuf-based protocol that allows multiple object types to be serialized in the same data stream. Simplifying a little bit, the client and server share a map that assigns an integer number to each query-able object type; this is usually done defining the `TypeId` in the `.proto` schema file. (see [link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/app.proto#L5) and [link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/review.proto#L5)). This `TypeId` is sent in front of each Protobuf message transmitted across the wire.
```

File: README.md [hunk 6]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
 
-You also need to implement a marshaller for your data model.
-That sounds scary but Protobuf gives you some parsers that really simplify the entire process.
-The main thing in the marshaller is the logic that selects the right parser for the type.
+On the client side the application must provide a Protostream marshaller for its own data types. That could sound scary, but Protobuf can do most of the hard work for this.
+A protostream essentially must:
+- set the correct `typeId` in the protostream message when marshalling data to the server ([see](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Marshaller.cs#L17-L24));
```
Comment (oraNod) [inline README.md]:
```suggestion
- Sets the correct `TypeId` in the ProtoStream message when marshalling data to the server ([see](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Marshaller.cs#L17-L24));
```

File: README.md [hunk 7]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
 
-You also need to implement a marshaller for your data model.
-That sounds scary but Protobuf gives you some parsers that really simplify the entire process.
-The main thing in the marshaller is the logic that selects the right parser for the type.
+On the client side the application must provide a Protostream marshaller for its own data types. That could sound scary, but Protobuf can do most of the hard work for this.
+A protostream essentially must:
```
Comment (oraNod) [inline README.md]:
```suggestion
A ProtoStream marshaller essentially does the following:
```

File: README.md [hunk 8]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
+- set the correct `typeId` in the protostream message when marshalling data to the server ([see](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Marshaller.cs#L17-L24));
+- use the correct Protobuf parser when unmarshalling from the server ([see](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Marshaller.cs#L31-L37)).
```
Comment (oraNod) [inline README.md]:
```suggestion
- Uses the correct Protobuf parser when unmarshalling objects from the server ([see](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Marshaller.cs#L31-L37)).
```

File: README.md [hunk 9]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
+Once the marshaller has been implemented, the plain C# classes generated by the Protobuf framework can be used in the application development.
 
-- In the Protobuf schema, a unique `TypeId` is assigned to each Protobuf message ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/app.proto#L5)) and ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/review.proto#L5)).
-- In the marshaller, the `TypeId` identifies which parser to select ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Marshaller.cs#L87)).
+Maybe it worth to highlight that, using protostream, the data format is independent from the specific programming language; this means that data store in the cache can be accessed by C#, Java, C++ clients. Actually the demo shows a round trip interoperability between different data representations:
```
Comment (oraNod) [inline README.md]:
```suggestion
It is worth highlighting that, using ProtoStream, the data format is independent from the specific programming language. This means that data stored in the cache can be accessed by C#, Java, C++ clients. Actually the demo shows round-trip interoperability between different data representations:
```

File: README.md [hunk 10]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
+- [json->C#](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L117) marshalling;
```
Comment (oraNod) [inline README.md]:
```suggestion
- [JSON->C#](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L117) marshalling.
```

File: README.md [hunk 11]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
+- [C#->protostream](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L118) marshalling;
```
Comment (oraNod) [inline README.md]:
```suggestion
- [C#->ProtoStream](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L118) marshalling.
```

File: README.md [hunk 12]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
+- [protostream->C#](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L46-L51) unmarshalling;
```
Comment (oraNod) [inline README.md]:
```suggestion
- [ProtoStream->C#](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L46-L51) unmarshalling.
```

File: README.md [hunk 13]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
+- [C#->json](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L52) marshalling.
```
Comment (oraNod) [inline README.md]:
```suggestion
- [C#->JSON](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L52) marshalling.
```

File: README.md [hunk 14]
```diff
@@ -61,20 +66,32 @@ podman stop query_demo
...
 
 ### Conclusion
 
-With the Infinispan Hot Rod .NET Core client it's easy for .NET applications to store and query remote caches while working with C# data types.
+With the Infinispan Hot Rod .NET Core client it's easy for .NET applications to store and query remote caches, in an language independent format, while working with native C# data types.
```
Comment (oraNod) [inline README.md]:
```suggestion
With the Infinispan Hot Rod .NET Core client it's easy for .NET applications to store and query remote caches, in an language neutral format, while working with native C# data types.
```

**Comment by rigazilla** [comment]
Applied all the @oraNod suggestions and merged, thanks!

**Comment by oraNod** [comment]
@rigazilla Nice one!

**Pull request merged**